File: Grind75_1/week6/BinaryTreeRightSideView.py
"""
https://leetcode.com/problems/binary-tree-right-side-view/description/
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        
        # BFS

        if (root is None):
            return []

        resList = []

        q = deque()

        q.append(root)

        arrLevelChild = [1]
        currLevel = 0
        
        while (len(q) != 0 and root is not None):

            for i in range(arrLevelChild[currLevel]):
                currNode = q.popleft()

                if (i == 0):
                    resList.append(currNode.val)

                if (currNode.right is not None):
                    q.append(currNode.right)
                if (currNode.left is not None):
                    q.append(currNode.left)

            arrLevelChild.append(len(q))
            currLevel = currLevel + 1

        return resList

            

File: Grind75/Binary-Tree/binary-tree-right-side-view.js
// https://leetcode.com/problems/binary-tree-right-side-view/description/

/**
BFS. level travsersal, go right subtree first and res.push the first value on the level

- Time: O(n)    // must still visit every node side the most right for a level can be not on the most right child
- Space: O(h)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    const res = new Array()
    const qu = new Deque()
    qu.pushBack(root)

    while (qu.size() > 0) {
        const quSize = qu.size()
        let right = true
        for (let i = 0; i < quSize; i ++) {
            const pop = qu.popFront()
            if (pop === null) {
                continue
            }

            if (right === true) {
                res.push(pop.val)
                right = false
            }

            if (pop.right !== null) {
                qu.pushBack(pop.right)
            }
            if (pop.left !== null) {
                qu.pushBack(pop.left)
            }
            
        }
    }

    return res
};